### ISSUE TITLE
[ENH] homogenize serialization design for `sktime` estimators with `skpro`

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#10582]
@patelchaitany has been working on a serialization design in https://github.com/sktime/skpro/pull/1073 in `skpro`.

This could in-principle be transferred to `sktime` - or at least we should check if it can be.

This issue is about:

* verifying how easy it is to transfer the `skpro` design to `sktime` - is it 1:1 adaptation, or are further refinements needed? This should be done first to estimate the amount of work needed to proceed.
* We should also review for compatibility with the deep learning serialization in #10453 resp #10450  by @geetu040 
* from this, derive suggested next steps to reach a state where `sktime` and `skpro` have compatible serialization and serialized data container design; where serialization works recursively across both repos; and where deep learning / foundation model estimators can be serialized too.

The "definition of done" for now is the first two bullet points: @geetu040 to verify whether this can be adapted, and estimating difficulty and work items for the homogenization (before starting with it).

Then we decide how we proceed, and who works on this (e.g., @patelchaitany or @geetu040 or someone else)

(issue for internal sktime devs only)

**Comment by patelchaitany** [comment]
I've looked at both PRs - my https://github.com/sktime/skpro/pull/1073 handles recursive decomposition of composite estimators, while #10453 handles native serialization of DL model weights. I think the way to merge them is to use recursive save for the tree structure, and at each leaf store artifact metadata (library, class, save/load method) in manifest.json so the core stays framework-agnostic. Estimators would declare their save/load recipe via tags, with a _restore_artifact() hook for framework-specific quirks - this should make save/load compatible across all estimator types.

wdyt @geetu040 , @fkiraly

**Comment by geetu040** [comment]
Copied from https://github.com/sktime/skpro/pull/1073#pullrequestreview-4743791360

FYI: @fkiraly @patelchaitany 

I think the design needs some changes, mostly how we format our dump folder. For this, I have left suggestions on the layout below.

### Current `sktime` layout

I have documented the current design in the PR description for https://github.com/sktime/sktime/pull/10453, but I'll add some for context here.

For an ordinary estimator, `sktime` keeps `_metadata` and `_obj` at the root:

```text
estimator.zip
├── _metadata
└── _obj
```

When the estimator owns native artifacts, those artifacts are stored alongside the regular object state:

```text
estimator.zip
├── _metadata
├── _obj
└── _artifacts/
    ├── index.json
    └── model_/
        ├── config.json
        └── model.safetensors
```

Another native backend might produce:

```text
estimator.zip
├── _metadata
├── _obj
└── _artifacts/
    ├── index.json
    └── network_/
        └── state_dict.pt
```

The important property is that a serialized estimator is a self-contained node:

```text
node
├── _metadata
├── _obj
└── _artifacts/       # optional
```

`_obj` contains the regular estimator state, while `_artifacts` contains attributes that must be saved using framework-native serialization.

### Current `skpro` layout

The recursive layout in this PR is different:

```text
estimator.zip
├── manifest.json
├── _format
├── _version
├── root/
│   ├── _metadata
│   └── _obj
└── components/
    ├── estimators__0__1/
    │   ├── _metadata
    │   └── _obj
    └── estimators__1__1/
        ├── _metadata
        └── _obj
```

I do not think the additional `root/` directory is needed. The archive root already represents the root estimator, so `_metadata` and `_obj` can remain there. This also preserves the existing `sktime` layout and makes a non-composite archive a natural subset of a composite archive.

### Proposed recursive layout

I suggest treating every estimator as the same self-contained serialization node and applying that node structure recursively:

```text
estimator.zip
├── _metadata
├── _obj
├── _artifacts/                   # optional native artifacts of root
│   ├── index.json
│   └── model_/
│       └── model.safetensors
└── _components/                  # optional child estimators
    ├── index.json
    ├── forecaster_/
    │   ├── _metadata
    │   ├── _obj
    │   ├── _artifacts/           # optional native artifacts of child
    │   │   ├── index.json
    │   │   └── network_/
    │   │       └── state_dict.pt
    │   └── _components/          # optional grandchildren
    │       ├── index.json
    │       └── transformer_/
    │           ├── _metadata
    │           └── _obj
    └── transformer_/
        ├── _metadata
        └── _obj
```

`_components/index.json` could describe how child nodes are connected to the parent:

```json
{
  "forecaster_": {
    "path": "forecaster_",
    "attribute": "forecaster_",
    "class": "sktime.forecasting.some_module.SomeForecaster"
  },
  "transformer_": {
    "path": "transformer_",
    "attribute": "transformer_",
    "class": "sktime.transformations.some_module.SomeTransformer"
  }
}
```

The parent `_obj` would contain placeholders or references for extracted child estimators. Loading would then:

1. Unpickle the parent `_obj`.
2. Restore the parent's native artifacts.
3. Read `_components/index.json`.
4. Recursively load each child using the same node contract.
5. Reattach each child at the recorded location.

This makes the format recursive without introducing a different serialization format for the root.

**Comment by fkiraly** [comment]
> Is there a design discussion or document that I am missing on, would be nice to share if there is any. For now I am relying on `serialization.rst` in `docs/` to understand the pattern.

There is not afaik, more precisely:

* there is a 3Y old step which comes close, but I think has not been perfectly implemented: https://github.com/sktime/enhancement-proposals/blob/main/steps/20_saveload/step.md
* the proposed `skl` time format has not been documented in the API documentation, as it should have been
* the current proposal is not entirely congruent with the old save/load proposal, so I think if we agree on the details it should be added as an enhancement-proposal, to supersede STEP 20.

For discussing the current proposal, I think we need to:

* define what `manifest.json` exactly contains (because it is present in @patelchaitany´s proposal but removed in @geetu040´s proposal, but neither of you seem to discuss what is in there or why it should be there, or why it should be removed)
* document the serialized format in a tentative PR to docs
* discuss how the serialized format diverges from current
* ultimately, summarize in a STEP

**Comment by geetu040** [comment]
In @srupat's implementation

generally it would look like this:
```json
{
  "version": 2,
  "serialization_format": "pickle",
  "root_class": "fully.qualified.RootClass",
  "is_fitted": true,
  "components": {},
  "load_order": []
}
```

and for example, the fitted `VotingProbaRegressor`, it would look approximately like this:
<details>

```json
{
  "version": 2,
  "serialization_format": "pickle",
  "root_class": "skpro.regression.ensemble._voting.VotingProbaRegressor",
  "is_fitted": true,
  "components": {
    "root": {
      "class": "skpro.regression.ensemble._voting.VotingProbaRegressor",
      "path": "root/",
      "is_leaf": false,
      "serialized_as": "structured",
      "children": [
        "estimators__0__1",
        "estimators__1__1",
        "estimators___0__1",
        "estimators___1__1"
      ],
      "parent": null,
      "attr_ref": ""
    },
    "estimators__0__1": {
      "class": "skpro.regression.residual.ResidualDouble",
      "path": "components/estimators__0__1/",
      "is_leaf": true,
      "serialized_as": "structured",
      "children": [],
      "parent": "root",
      "attr_ref": "estimators[0][1]"
    },
    "estimators__1__1": {
      "class": "skpro.regression.residual.ResidualDouble",
      "path": "components/estimators__1__1/",
      "is_leaf": true,
      "serialized_as": "structured",
      "children": [],
      "parent": "root",
      "attr_ref": "estimators[1][1]"
    },
    "estimators___0__1": {
      "class": "skpro.regression.residual.ResidualDouble",
      "path": "components/estimators___0__1/",
      "is_leaf": true,
      "serialized_as": "structured",
      "children": [],
      "parent": "root",
      "attr_ref": "estimators_[0][1]"
    },
    "estimators___1__1": {
      "class": "skpro.regression.residual.ResidualDouble",
      "path": "components/estimators___1__1/",
      "is_leaf": true,
      "serialized_as": "structured",
      "children": [],
      "parent": "root",
      "attr_ref": "estimators_[1][1]"
    }
  },
  "load_order": [
    "estimators__0__1",
    "estimators__1__1",
    "estimators___0__1",
    "estimators___1__1",
    "root"
  ]
}
```

</details>

Here I am suggesting to move `version`, `serialization_format`, `class` to `_metadata`
And put mapping of variable name to dumped path in `index.json` for all components or avoid this and keep the respective folder name accordingly.

**Comment by GunaTeja777** [comment]
Thanks @fkiraly! Understood. I'll follow the discussion in #10582 / #10584 and update this PR to align with the new serialization layout once it's finalized.

**Comment by patelchaitany** [comment]
@geetu040 Personally I'm not a big fan of the deep recursive tree - I'd prefer a flat index with one big manifest.json that contains all the information about saved components. The recursive tree structure has its benefit of not needing an explicit load order, but with a flat layout the index.json basically becomes part of manifest.json. For loading, we first restore the leaves, then pass the relevant info (paths, metadata) up to the root class, which calls each child's native loading framework with the necessary parameters.

**Comment by fkiraly** [comment]
@patelchaitany, interesting idea. We could follow an indexing structure with flat structure but nesting is implicit, not explicit. However, that seems harder to keep track of and implement.

My gut feeling is that the explicit nesting structure is easier to look at directly (e.g., a developer unpacking the zip), and easier to get an intellectual grip of. This is a big advantage, of course we need to compare more pros/cons to alternatives in our discussion.

**Comment by geetu040** [comment]
@fkiraly @patelchaitany coming on the topic of compatibility between native serialization and the recursive serialization design

From the issue description:
> * We should also review for compatibility with the deep learning serialization in [[ENH] Native Serialization for Deep Learning Estimators #10453](https://github.com/sktime/sktime/pull/10453) resp [[ENH] Native serialization for deep learning estimators #10450](https://github.com/sktime/sktime/issues/10450)  by [@geetu040](https://github.com/geetu040)

I think the native serialization is compatible with the current design, my proposed design and probably any design. Because the current native serialization serializes only the relevant attributes at the node-level. If you take a look at these lines [sktime/base/_serialize.py#L435-L444](https://github.com/geetu040/sktime/blob/f7665f119705f7ea43da588aa669da5dc6cee6ca/sktime/base/_serialize.py#L435-L444) in #10453. You will notice that native artifacts are handled in the same code block where we handle `_metadata` and `_obj`. So we are hopefully safe here.

---

To be decided is our recursive serialization design which is being implemented in https://github.com/sktime/skpro/pull/1073 and will be used for sktime later.

**Comment by patelchaitany** [comment]
> My gut feeling is that the explicit nesting structure is easier to look at directly (e.g., a developer unpacking the zip), and easier to get an intellectual grip of. This is a big advantage, of course we need to compare more pros/cons to alternatives in our discussion.

i think it is easier to look and Implementing it would be also easier.
